### PR TITLE
[AMD] [DSV4]:  fp8 target lm_head GEMM for DSPARK verify logits

### PR DESCRIPTION
**Pull request by karverma-amd**
## Summary

On the DeepSeek-V4 DSPARK speculative-decode path, the target `lm_head` projection that produces the **verify logits** runs a bf16 matmul `[num_verify_tokens, hidden] x [hidden, vocab_slice]` in `project_through_lm_head`. This is a per-decode-step cost that scales with the verify batch (≈ `concurrency × (gamma+1)` tokens) and, on the trace, dispatches to a `Cijk_*` (rocBLAS/Tensile) bf16 GEMM. This PR adds an **opt-in fp8 path** for that GEMM (`torch._scaled_mm`, per-token activation scale + per-output-channel weight scale, weight quantized once and cached on the module), behind a new env flag, with the bf16 matmul as fallback.

**What this change is:** an opt-in fp8 replacement of the DSPARK verify-head GEMM. Off by default, it roughly halves that one kernel; end-to-end it is a **small, decode-bound win (~+1–1.5% tput / −3% TPOT at high concurrency)** with **no GSM8K accuracy regression**.

## Motivation

Attributing the DSV4 DSPARK decode step kernel-by-kernel (torch-profiler) shows the target-head GEMM that produces the `[num_verify_tokens, vocab]` logits is one of the per-step GEMMs, and it grows with the verify batch. It is a plain bf16 `[M, K] x [K, N]` with a very large `N` (vocab) and small `M`, so it is weight-read (memory) bound — halving the weight bytes in fp8 nearly halves the kernel. This is the last per-step decode GEMM on this path not already running in a reduced precision.

## What changed

* `dspark.py`: two module helpers — `_fp8_quant_per_row(x)` (per-row symmetric fp8 quant) and `_fp8_lm_head_matmul(hidden, lm_head)` (`hidden @ weight.T` via `torch._scaled_mm`, caching the fp8 weight + scale on the `lm_head` module). `project_through_lm_head` calls the fp8 path when the flag is on, the head is floating-point (non-quantized), `hidden.shape[0] > 0`, and the vocab slice is 16-aligned; otherwise it uses the original `torch.matmul`. A quantized head keeps its own kernel; the tiny `M=1` sample head stays bf16.
* `environ.py`: new `SGLANG_OPT_DSPARK_FP8_LM_HEAD = EnvBool(False)`.
* The flag is read **once at module import** (`_DSPARK_FP8_LM_HEAD`) so the per-token decode path has no env lookup.

No behavior change unless the flag is set.

## Validation (MI35X, DeepSeek-V4-Pro-0813 FP4, TP8 + DP-attention, DSPARK gamma=6)

**Numerics** — model-free parity check of `torch._scaled_mm` (per-row act + per-channel weight fp8) vs the bf16 matmul on this stack (gfx950): mean abs error ~0.076, p99 relative error ~2.4% (fp8 rounding); output magnitudes match. The precision loss is bounded and does not move greedy argmax on real (peaked) LM logits — see accuracy.

**Accuracy** — 5-shot GSM8K, 1319 questions, `--parallel 1319`, greedy (temp 0), cuda-graph on, **real speculative acceptance** (`SGLANG_SIMULATE_ACC_*` unset so outputs are true model outputs):

| lm_head dtype | Accuracy | Invalid |
| --- | --- | --- |
| bf16 (flag off) | 0.939 | 0.000 |
| fp8 (flag on) | 0.948 | 0.000 |

No regression — the +0.9pt is within greedy-divergence run-to-run noise, 0% invalid on both.

**Kernel swap (trace-confirmed)** — profiled TP8, conc16, rank-0, over the decode window. The verify-head GEMM flips 1:1 between arms and nothing else changes:

* off: `Cijk_Alik_Bljk_BBS_..._MT512x16x64` (bf16) — ~0.567 ms/step (2 launches/step)
* on:  `Cijk_Alik_Bljk_F8BS_..._MT256x16x256` (fp8) — ~0.153 ms/step (1 launch/step)

≈73% cheaper on that GEMM. Amortized over the accepted-token count this reconciles with the measured +1.1% at conc16.

**Throughput** (ISL 8192 / OSL 1024, decode-inclusive shape; flag off vs on):

| conc | metric | off (bf16) | on (fp8) | Δ |
| --- | --- | --- | --- | --- |
| 16 | median TPOT | 9.58 ms | 9.57 ms | ~flat |
| 16 | decode tok/s | 777.5 | 785.8 | +1.1% |
| 128 | median TPOT | 25.58 ms | 24.80 ms | **−3.0%** |
| 128 | decode tok/s | 2670.5 | 2707.2 | **+1.4%** |

Honest framing: a small, decode-bound win that grows with the verify batch (bigger `M` at higher concurrency). Under prefill-dominated shapes it is ~neutral.

## How to enable

```
SGLANG_OPT_DSPARK_FP8_LM_HEAD=1
```

Applies to the DSPARK target verify head only; requires a floating-point (non-quantized) `lm_head`. Off by default.

## Checklist

* Opt-in (default off); quantized head, `M=1` sample head, and the bf16 fallback are unchanged
* Numerically validated (bounded fp8 error; GSM8K accuracy preserved on real acceptance)
* Kernel swap trace-confirmed (bf16 `Cijk_*` → fp8 `Cijk_*_F8BS_*`, 1:1)
* Repeated A/B throughput (TPOT + decode tok/s) at conc 16 / 128
* Flag read once at module import (no per-token env lookup)

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #35656133545](https://github.com/sgl-project/sglang/actions/runs/35656133545)
Latest PR Test (Extra): :x: [Run #35656133243](https://github.com/sgl-project/sglang/actions/runs/35656133243)
Latest PR Test (AMD ROCm 10): :x: [Run #35656133494](https://github.com/sgl-project/sglang/actions/runs/35656133494)

### CODE REVIEW

File: python/sglang/srt/models/dspark.py [hunk 1]
```diff
@@ -28,6 +28,47 @@
...
+# fp8 target lm_head GEMM for the DSPARK verify logits, opt-in via
+# SGLANG_OPT_DSPARK_FP8_LM_HEAD. The bf16 head matmul
+# [num_verify_tokens, hidden] x [hidden, vocab_slice] is a per-decode-step cost
+# that grows with the verify batch; running it in fp8 (per-token activation
+# scale + per-output-channel weight scale) roughly halves that GEMM. The weight
+# is fp8-quantized once and cached on the module. Evaluated once at import to
+# keep the per-token decode path free of env lookups.
+_DSPARK_FP8_LM_HEAD = envs.SGLANG_OPT_DSPARK_FP8_LM_HEAD.get()
+_FP8_DTYPE = torch.float8_e4m3fn
+_FP8_MAX = 448.0
```
Comment (kkHuang-amd) [inline python/sglang/srt/models/dspark.py]: Could you add platform flag to protect it?

If user ran it with SGLANG_OPT_DSPARK_FP8_LM_HEAD on gfx942, it will have acc issue.

File: python/sglang/srt/models/dspark.py [hunk 2]
```diff
@@ -28,6 +28,47 @@
...
+    h_fp8, h_scale = _fp8_quant_per_row(hidden)  # [M,K], [M,1]
+    # a=[M,K] fp8, b=[K,N] fp8 (col-major via .t()); scale_a per-row [M,1],
+    # scale_b per-col [1,N]. Output cast back to hidden dtype.
+    return torch._scaled_mm(
+        h_fp8,
+        w_fp8.t(),
+        scale_a=h_scale,
+        scale_b=w_scale.t(),
+        out_dtype=hidden.dtype,
+    )
```
Comment (kkHuang-amd) [inline python/sglang/srt/models/dspark.py]: Have you ever tried to use the aiter tgemm to replace torch._scaled_mm?

**Comment by karverma-amd** [comment]
Thanks @kkHuang-amd — good catch. Added a platform guard in `3897074`: the flag is now `envs.SGLANG_OPT_DSPARK_FP8_LM_HEAD.get() and is_gfx95_supported()`, so the fp8 `_scaled_mm` head path only engages on gfx95x (where it's validated). On gfx942 the flag stays inert and the original bf16 matmul is used, so there's no accuracy regression; a one-time warning is logged if the env is set on an unsupported arch.

File: python/sglang/srt/models/dspark.py [hunk 3]
```diff
@@ -25,9 +25,61 @@
...
+    w_fp8 = getattr(lm_head, "_fp8_weight_cache", None)
+    w_scale = getattr(lm_head, "_fp8_wscale_cache", None)
+    if w_fp8 is None or w_fp8.shape != weight.shape:
+        w_fp8, w_scale = _fp8_quant_per_row(weight)  # [N,K], [N,1]
+        lm_head._fp8_weight_cache = w_fp8
+        lm_head._fp8_wscale_cache = w_scale
+    h_fp8, h_scale = _fp8_quant_per_row(hidden)  # [M,K], [M,1]
+    # a=[M,K] fp8, b=[K,N] fp8 (col-major via .t()); scale_a per-row [M,1],
+    # scale_b per-col [1,N]. Output cast back to hidden dtype.
+    return torch._scaled_mm(
```
Comment (kkHuang-amd) [inline python/sglang/srt/models/dspark.py:74]: Have you ever tried to use the aiter tgemm to replace torch._scaled_mm?

Comment (karverma-amd) [reply python/sglang/srt/models/dspark.py:74]: @kkHuang-amd : I dug in on the actual verify-head shape ([M, 7168] × [7168, vocab≈129280], fp8 e4m3fn, gfx950):

aiter.tgemm (TunedGemm) is the bf16 tuned-GEMM — fp8 inputs fall through to addmm and error, so it can't replace _scaled_mm.
aiter.gemm_a8w8 (the real per-token fp8 GEMM) is numerically identical (cos 0.9993) but untuned it's 12–61% slower. So I tuned it (gemm_a8w8_tune.py) at the padded-M grid for this shape and compared the winning kernels vs _scaled_mm: tied at M≤32, ~4–6% slower at M=64–128, ~23% slower at M=256. It regresses exactly where the verify batch is largest (high concurrency).
Both fp8 paths are ~2× over bf16, so _scaled_mm already captures the win, and even a tuned gemm_a8w8 doesn't beat it on this large-N/weight-read-bound GEMM. So I'll keep torch._scaled_mm
image

**Review by kkHuang-amd — APPROVED** [review]
LGTM
